SSTOOLIKIT-75 (#68)
nordic-institute/X-Road-Security-Server-toolkit: docs/xroad_security_server_toolkit_user_guide.md

| 24.05.2021 | 1.3.7       | Added download-internal-tsl command                                          | Alberto Fernandez  |
| 28.05.2021 | 1.3.8       | Added member name property  and multitenancy section                         | Alberto Fernandez  |
| 28.05.2021 | 1.3.9       | Update service management                                                    | Bert Viikmäe       |

> | 04.06.2021 | 1.3.8       | Refactor documentation                                                       | Alberto Fernandez  |

Reviewer: The version number should increment

Reviewer: Note that the version number is also stated at the top of the document

			    - <SERVICE_CLIENT_ID>
```

#### 3.2.1 Access Configuration

> In this section is configured the security server access as described in

Reviewer: ```suggestion
This section shows how to configure the security server access as described in
```

  used for all configurable security servers, but if specified in the ``security_server`` section, the value will be overridden for specific 
  configurable security server)
* <SSH_USER_OS_ENV_VAR_NAME> Environment variable name to hold SSH username, e.g. if the variable is set like ``export TOOLKIT_SSH_USER=ssh_user`` the value to use here is ``TOOLKIT_SSH_USER`` (if specified in ``ssh_access`` section, one value will be used for all configurable security servers, 
  but if specified in the ``security_server`` section, the value will be overridden for specific configurable security server)
* <SSH_PRIVATE_KEY_OS_ENV_VAR_NAME> Environment variable name to hold full path to SSH private key, e.g. if the variable is set like ``export TOOLKIT_SSH_PRIVATE_KEY=/home/user/private_key`` the value to use here is ``TOOLKIT_SSH_PRIVATE_KEY``
  (if specified in ``ssh_access`` section, one value will be used for all configurable security servers, 
  but if specified in the ``security_server`` section, the value will be overridden for specific configurable security server)    

#### 3.2.2 Security Servers Configuration

> In this section is configured the security servers information. It is possible to set multiple security servers that will run in a sequential batch mode.

Reviewer: ```suggestion
This section shows how to set up the security server information. It is possible to configure multiple security servers at the same time. The toolkit will execute the configurations sequentially.
```

```
security_server:
- api_key: <API_KEY_ENV_VAR_NAME>
  api_key_url: https://localhost:4000/api/v1/api-keys
  admin_credentials: <SECURITY_SERVER_CREDENTIALS_OS_ENV_VAR_NAME>
  configuration_anchor: <CONFIGURATION_ANCHOR_PATH>
  certificates:
    - <SIGN_CERT_PATH>
    - <AUTH_CERT_PATH>
  name: <SECURITY_SERVER_NAME>
  owner_dn_country: <OWNER_DISTINGUISHED_NAME_COUNTRY>
  owner_dn_org: <OWNER_DISTINGUISHED_NAME_ORGANIZATION>
  owner_member_class: <OWNER_MEMBER_CLASS>
  owner_member_code: <OWNER_MEMBER_CODE>
  security_server_code: <SERVER_CODE>
  software_token_id: <SOFT_TOKEN_ID>
  software_token_pin: <SOFT_TOKEN_PIN>
  fqdn: <SECURITY_SERVER_EXTERNAL_FQDN>
  url: https://<SECURITY_SERVER_INTERNAL_FQDN_OR_IP>:4000/api/v1
  ssh_user: <SSH_USER_OS_ENV_VAR_NAME>
  ssh_private_key: <SSH_PRIVATE_KEY_OS_ENV_VAR_NAME>
```
* <API_KEY_ENV_VAR_NAME> Environment variable name to hold X-Road Security Server API key (e.g. if the variable is set like ``export TOOLKIT_API_KEY=f13d5108-7799-426d-a024-1300f52f4a51`` the value to use here is ``TOOLKIT_API_KEY``) or left as-is/any for toolkit to attempt creation of transient API key

> * <SECURITY_SERVER_CREDENTIALS_OS_ENV_VAR_NAME> (Optional) If is set it will overwritte the <SECURITY_SERVER_CREDENTIALS_OS_ENV_VAR_NAME> property described in the [access section](#3.2.1-access-configuration)

Reviewer: ```suggestion
* <SECURITY_SERVER_CREDENTIALS_OS_ENV_VAR_NAME> (Optional) If is set it will overwrite the <SECURITY_SERVER_CREDENTIALS_OS_ENV_VAR_NAME> property described in the [access section](#3.2.1-access-configuration)
```

* <CONFIGURATION_ANCHOR_PATH> Path to the configuration anchor file, e.g. "/etc/xroad/configuration-anchor.xml"
*  <SIGN_CERT_PATH> should be given as path referring to sign certificates location.
*  <AUTH_CERT_PATH> should be given as path referring to auth certificate location.
* <SECURITY_SERVER_NAME> should be substituted with the installed security server name, e.g. ss1
* <OWNER_DISTINGUISHED_NAME_COUNTRY> should be ISO 3166-1 alpha-2 two letter code for server owner country. This is used in certificate generation.
* <OWNER_DISTINGUISHED_NAME_ORGANIZATION> should be set to server owner organization. This is used in certificate generation.
* <OWNER_MEMBER_CLASS> should be substituted with the member class obtained from the Central Server, e.g. GOV.
* <OWNER_MEMBER_CODE> should be substituted with the member code obtained from the Central Server, e.g. 1234.
* <SERVER_CODE> should be substituted with the server code of the installed security server, e.g. SS1
* <SOFT_TOKEN_ID> default software token ID, normally 0 (zero).
* <SOFT_TOKEN_PIN> should be substituted with a desired numeric pin code.

> * <SECURITY_SERVER_EXTERNAL_FQDN> externally accessible FQDN for security server, propagates to security server certificates.

Reviewer: ```suggestion
* <SECURITY_SERVER_EXTERNAL_FQDN> externally accessible FQDN for the security server. It's applied to the security server certificates.
```

* <SECURITY_SERVER_INTERNAL_FQDN_OR_IP> should be substituted with internal IP address or host name of the installed security server, e.g. ``ss1``

> * <SSH_USER_OS_ENV_VAR_NAME> (Optional) If is set it will overwritte the <SSH_USER_OS_ENV_VAR_NAME> property described in the [access section](#3.2.1-access-configuration)

Reviewer: ```suggestion
* <SSH_USER_OS_ENV_VAR_NAME> (Optional) If set, it will overwrite the <SSH_USER_OS_ENV_VAR_NAME> property described in the [access section](#3.2.1-access-configuration)
```

> * <SSH_PRIVATE_KEY_OS_ENV_VAR_NAME> (Optional) If is set it will overwritte the <SSH_PRIVATE_KEY_OS_ENV_VAR_NAME> property described in the [access section](#3.2.1-access-configuration)

Reviewer: ```suggestion
* <SSH_PRIVATE_KEY_OS_ENV_VAR_NAME> (Optional) If set, it will overwrite the <SSH_PRIVATE_KEY_OS_ENV_VAR_NAME> property described in the [access section](#3.2.1-access-configuration)
```

> #### 3.2.3 Clients Servers Configuration

Reviewer: Perhaps just `Client Configuration`?

> In this section is configured the clients information. It is possible to set a list of subsystems belonging to the owner member

Reviewer: ```suggestion
The security server client information is configured in this section. It is possible to set up a list of subsystems belonging to the owner member
```

> configured in the [security server configuration](#322-security-servers-configuration) or it's possible to add a new member with its subsystems as described

Reviewer: ```suggestion
configured in the [security server configuration](#322-security-servers-configuration) or add them to a new member as described
```

in [8 Multitenancy](#8-multitenancy).
```
clients:
	- member_class: <MEMBER_CLASS>
	  member_code: <MEMBER_CODE>
	  member_name: <MEMBER_NAME>
	  subsystem_code: <SUBSYSTEM_CODE>
	  connection_type: <CONNECTION_TYPE>
```

* <MEMBER_CLASS> should be substituted with the member class obtained from the Central Server, e.g. GOV.
It must have the same value as <OWNER_MEMBER_CLASS> if is a subsystem of the owner client.
* <OWNER_MEMBER_CODE> should be substituted with the member code obtained from the Central Server, e.g. 1234
It must have the same value as <OWNER_MEMBER_CLASS> if is a subsystem of the owner client.
* <MEMBER_NAME> should be substituted with the member name obtained from the Central Server, e.g. COMPANY.
It must have the same value as <OWNER_DISTINGUISHED_NAME_ORGANIZATION>  if is a subsystem of the owner client.
* <SUBSYSTEM_CODE> X-Road member/client subsystem code.
* <CONNECTION_TYPE> Connection protocol selection, from among ``HTTP``, ``HTTPS``, ``HTTPS_NO_AUTH``.

> #### 3.2.3 Services Configuration

Reviewer: ```suggestion
#### 3.2.3 Service Configuration
```

```
service_descriptions:
	- url: <SERVICE_DESCRIPTION_URL>
	  rest_service_code: <REST_SERVICE_CODE>
	  type: <SERVICE_TYPE>
	  access:
		- <SERVICE_DESCRIPTION_ACCESS>
	  url_all: <SERVICE_URL_FOR_ALL>
	  timeout_all: <SERVICE_TIMEOUT_FOR_ALL>
	  ssl_auth_all: <SERVICE_USE_SSL_AUTH_FOR_ALL>
	  services:
            - service_code: <SERVICE_CODE>
              access:
                - <SERVICE_ACCESS>
              timeout: <SERVICE_TIMEOUT>
              ssl_auth: <SERVICE_USE_SSL_AUTH>
              url: <SERVICE_URL>
          endpoints:
			- path: <ENDPOINT_PATH>
			  method: <ENDPOINT_METHOD>
			  access: 
			    - <ENDPOINTS_ACCESS>
```
<strong>Service description:</strong>

* <SERVICE_DESCRIPTION_URL> URL for service description.
* <REST_SERVICE_CODE> rest service code, not used for WSDL services.
* <SERVICE_TYPE> type of service, value from ``OPENAPI3``, ``REST``, ``WSDL``.
* <SERVICE_DESCRIPTION_ACCESS> (Optional) list of subsystems ids, composed by <INSTANCE>:<MEMBER_CLASS>:<MEMBER_CODE>:<SUBSYSTEM_CODE>, or security serverver owners composed 
by <INSTANCE>:security-server-owners.

> * <SERVICE_URL_FOR_ALL> boolean value determining URL prefix should be the same for all services.

Reviewer: ```suggestion
* <SERVICE_URL_FOR_ALL> boolean value determining if the URL prefix should be applied for all the services.
```

> * <SERVICE_TIMEOUT_FOR_ALL> boolean value specifying timeout should be the same for all services in service description.

Reviewer: ```suggestion
* <SERVICE_TIMEOUT_FOR_ALL> boolean value specifying if the timeout should be applied for all the services in the service description.
```

> * <SERVICE_USE_SSL_AUTH_FOR_ALL> boolean value for specifying whether SSL authentication should be used for all services for a given service description.

Reviewer: ```suggestion
* <SERVICE_USE_SSL_AUTH_FOR_ALL> boolean value specifying whether SSL authentication should be used for all the services in the service description.
```

<strong>Services (Optional):</strong>
The services are autogenerated when the service description is added, therefore it is not necessary to add all the services for the service description url, 
we only have to add the services when we want to customize the access rights, update the parameters (of all or just one) or create new services 

> that are not self-generated.

Reviewer: ```suggestion
It's possible to create services manually but usually it's not required because they are parsed from the service descriptions automatically. Adding a service manually is needed when a service description isn't available, there is a need to customize the access rights, or there is a need to customize the parameters.
```

* <SERVICE_CODE> code for single service.
* <SERVICE_ACCESS> (Optional) same as <SERVICE_DESCRIPTION_ACCESS>, if its defined, it will overwrite the values defined in <SERVICE_DESCRIPTION_ACCESS>
for the single service.
* <SERVICE_TIMEOUT> timeout for service in seconds.
* <SERVICE_USE_SSL_AUTH> boolean value for specifying whether SSL authentication should be used.
* <SERVICE_URL> URL for single service

> <strong>Endpoints (Optional):</strong>

Reviewer: Maybe we should also say that endpoints are only relevant to REST services

The whole security server configuration in a fully automatic mode (all configuration from configuration file) can be run with ``xrdsst apply``
For performing the configuration step by step instead, please start from [4.3 Initializing the security server](#43-initializing-the-security-server)

### 4.2 X-Road Security Server  Toolkit commands

> * Access rightss: All commands require at least the role XROAD_SYSTEM_ADMINISTRATOR.

Reviewer: ```suggestion
* Access rights: All commands require at least the role XROAD_SYSTEM_ADMINISTRATOR.
```

#### 4.2.1 Creating admin user command

X-Road admin user can be created with 
```
xrdsst user create-admin
```

> Configuration parameters involved are the described in [3.2.1 Access Configuration](#321-access-configuration), 
> specifically, the administrator user will have the name: password those defined in the property admin_credentials 

Reviewer: ```suggestion
Configuration parameters involved are described in [3.2.1 Access Configuration](#321-access-configuration). Specifically, the administrator user will have the `name: password` defined in the property `admin_credentials`.
```

Note: This is an optional step in the configuration process and should only be run if the admin user has not been created before. 
      SSH (SSH user and a private key) is used for creating the admin user. If the admin user has been created before, then it is 
      enough to just add the credentials to the configuration file as ``admin_credentials`` and this step can be skipped totally.

> ** It is a security risk to store the SSH access related credentials into to configuration file as plain text. **

Reviewer: ```suggestion
**It is a security risk to store the SSH access-related credentials into to configuration file as plain text.**
```
